src/spikeinterface/core/sorting_tools.py

```diff
@@ -0,0 +1,90 @@
+import numpy as np
+
+
+def spike_vector_to_spike_trains(spike_vector: list[np.array], unit_ids: np.array) -> dict[dict]:
+    """
+    Computes all spike trains for all units/segments from a spike vector list.
+
+    Internally calls numba if numba is installed.
+
+    Parameters
+    ----------
+    spike_vector: list[np.ndarray]
+        List of spike vectors optained with sorting.to_spike_vector(concatenated=False)
+    unit_ids: np.array
+        Unit ids
+
+    Returns
+    -------
+    spike_trains: dict[dict]:
+        A dict containing, for each segment, the spike trains of all units
+        (as a dict: unit_id --> spike_train).
+    """
+
+    try:
+        import numba
+
+        HAVE_NUMBA = True
+    except:
+        HAVE_NUMBA = False
+
+    if HAVE_NUMBA:
+        # the trick here is to have a function getter
+        vector_to_list_of_spiketrain = get_numba_vector_to_list_of_spiketrain()
+    else:
+        vector_to_list_of_spiketrain = vector_to_list_of_spiketrain_numpy
+
+    num_units = unit_ids.size
+    spike_trains = {}
+    for segment_index, spikes in enumerate(spike_vector):
+        sample_indices = np.array(spikes["sample_index"]).astype(np.int64, copy=False)
+        unit_indices = np.array(spikes["unit_index"]).astype(np.int64, copy=False)
+        list_of_spiketrains = vector_to_list_of_spiketrain(sample_indices, unit_indices, num_units)
+        spike_trains[segment_index] = dict(zip(unit_ids, list_of_spiketrains))
+
+    return spike_trains
+
+
+def vector_to_list_of_spiketrain_numpy(sample_indices, unit_indices, num_units):
```

Comment:
If the "old way" you mean using the numpy version then that is what is super-slow and I think should be avoided.

@samuelgarcia @DradeAW 

If by the "old way" you mean just calling get_unit_spike_train for every unit (which is what I meant) then I think is indeed a bug because is calling this function.

Reply:
In my version, it was calling `get_unit_spike_train(unit_id)`.

But the function Sam wrote should be doing the same thing the `NumpySorting` does when you call `get_unit_spike_train` ... So it should be the same performance

Reply:
If the sorting is a NumpySorting that would be correct, and it makes sense: NumpySorting is slow at extracting spike trains because, wel, the representation is bad for that. That's what it is. 

But if you sorting is not as bad as Numpy for calling `get_unit_spike_train` then the current implementation is terrible without Numba. 

That is, If you have a `spike_vector` cached in something that is not as bad `NumpySorting` for extracting spikes *and* you don't have Numba then you will be silently calling a terribly slow function.

Reply:
Ah I see the problem!

You are speaking about a (corner) case where the `get_unit_spike_train` function is very fast, yet for some reason the spike trains are not cached but the spike vector is.
Indeed you are correct, in this case it would be terribly slow.

Reply:
I take issue with you calling that `corner_case`. Why are you so confident than sorting to build the spike_vector and then masking within resulting structured array is a faster operation than just calling `get_unit_spike_train`? 

My original appeal was to restrict this to methods that already have the `spike_vector` pre-calculated like `NumpySorting` and.a as  you were saying, maybe Phy. There, you can be confident than calling this unlikely to be bad. But then I tested this method with numba and it works OK even for those cases with kind of fast-io so that seems OK.

But for functions that that don't have `spike_vector` pre-calculated this seems like a corner case:

1) You already called `get_unit_spike_train` to get your `spike_vector` (so your data should be cached already, not need for this)
2) Then, you already called `get_unit_spike_train` but somehow it is better to mask a very long vector (generating a spike in memory) and extract it from there.

So again, I think this works great for cases that already have `spike_vector` pre-calculated and it works OK even for other cases (if you have numba), but, if you don't have it? or if you don't have numba? Are you sure that `get_unit_spike_train` is that bad in general?

Reply:
Ah I read again the function, and realized it does not check whether the spike trains cache is already full (this is in another PR). This should be added.

The corner case I was talking about is when `get_unit_spike_train` is very fast, yet the spike vector is cached but not the spike trains (since the spike vector is computed form the spike trains, they should be cached).